pkg/blockbuilder/scheduler/queue.go

```diff
@@ -3,123 +3,149 @@ package scheduler
+			return a.Priority > b.Priority // Higher priority first
+		}),
+		inProgress: make(map[string]*inProgressJob),
+		completed:  NewCircularBuffer[*types.Job](defaultCompletedJobsCapacity),
+		statusMap:  make(map[string]types.JobStatus),
 	}
 }
 
 func (q *JobQueue) Exists(job *types.Job) (types.JobStatus, bool) {
 	q.mu.RLock()
 	defer q.mu.RUnlock()
 
-	if _, ok := q.inProgress[job.ID]; ok {
-		return types.JobStatusInProgress, true
-	}
-
-	if _, ok := q.pending[job.ID]; ok {
-		return types.JobStatusPending, true
-	}
-
-	if _, ok := q.completed[job.ID]; ok {
-		return types.JobStatusComplete, true
-	}
-
-	return -1, false
+	status, exists := q.statusMap[job.ID]
+	return status, exists
 }
 
-// Enqueue adds a new job to the pending queue
-// This is a naive implementation, intended to be refactored
-func (q *JobQueue) Enqueue(job *types.Job) error {
+// Enqueue adds a new job to the pending queue with a priority
+func (q *JobQueue) Enqueue(job *types.Job, priority int) error {
 	q.mu.Lock()
 	defer q.mu.Unlock()
 
-	if _, exists := q.pending[job.ID]; exists {
-		return fmt.Errorf("job %s already exists in pending queue", job.ID)
-	}
-	if _, exists := q.inProgress[job.ID]; exists {
-		return fmt.Errorf("job %s already exists in progress", job.ID)
-	}
-	if _, exists := q.completed[job.ID]; exists {
-		return fmt.Errorf("job %s already completed", job.ID)
+	// Check if job already exists
+	if status, exists := q.statusMap[job.ID]; exists {
+		return fmt.Errorf("job %s already exists with status %v", job.ID, status)
 	}
 
-	q.pending[job.ID] = job
+	jobWithPriority := NewJobWithPriority(job, priority)
+	q.pending.Push(jobWithPriority)
+	q.statusMap[job.ID] = types.JobStatusPending
 	return nil
 }
 
 // Dequeue gets the next available job and assigns it to a builder
-func (q *JobQueue) Dequeue(builderID string) (*types.Job, bool, error) {
+func (q *JobQueue) Dequeue(_ string) (*types.Job, bool, error) {
 	q.mu.Lock()
 	defer q.mu.Unlock()
 
-	// Simple FIFO for now
-	for id, job := range q.pending {
-		delete(q.pending, id)
-		q.inProgress[id] = &jobAssignment{
-			job:       job,
-			builderID: builderID,
-		}
-		return job, true, nil
+	if q.pending.Len() == 0 {
+		return nil, false, nil
 	}
 
-	return nil, false, nil
+	jobWithPriority, ok := q.pending.Pop()
+	if !ok {
+		return nil, false, nil
+	}
+
+	// Add to in-progress with current time
+	q.inProgress[jobWithPriority.Job.ID] = &inProgressJob{
+		job:       jobWithPriority.Job,
+		startTime: time.Now(),
+	}
+	q.statusMap[jobWithPriority.Job.ID] = types.JobStatusInProgress
+
+	return jobWithPriority.Job, true, nil
 }
 
 // MarkComplete moves a job from in-progress to completed
-func (q *JobQueue) MarkComplete(jobID string, builderID string) error {
+func (q *JobQueue) MarkComplete(jobID string) error {
 	q.mu.Lock()
 	defer q.mu.Unlock()
 
-	assignment, exists := q.inProgress[jobID]
+	// Find job in in-progress map
+	inProgressJob, exists := q.inProgress[jobID]
```

Comment:
nit: maybe this should not error? if a scheduler restarted and lost it's state we'd miss committing already consumed jobs.

but this can be worried about later when we add the committer logic